src/pip/_internal/req/constructors.py

```diff
@@ -334,6 +334,13 @@ def install_req_from_req_string(
             "%s depends on %s " % (comes_from.name, req)
         )
 
+    if req.url:
+        # Create an InstallRequirement for a PEP 508 URL with the same behavior
+        # as 'pip install req.url'
+        return install_req_from_line(
+            req.url, comes_from=comes_from, isolated=isolated,
```

Comment:
What about extras?

PEP 508 allows them: `url_req = name wsp* extras? wsp* urlspec wsp+ quoted_marker?`

`pip install requests[security] @ https://files.pythonhosted.org/packages/52/2c/514e4ac25da2b08ca5a464c50463682126385c4272c18193876e91f4bc38/requests-2.21.0.tar.gz`

Reply:
oooh good catch. I'll try figuring that out...

> `pip install requests[security] @ https://files.pythonhosted.org/packages/52/2c/514e4ac25da2b08ca5a464c50463682126385c4272c18193876e91f4bc38/requests-2.21.0.tar.gz`

I think this actually doesn't work with `pip install`:

```console
$ pip --version
pip 19.0.3 from /Users/andy/.virtualenvs/tmp-de2709e9577734b/lib/python3.7/site-packages/pip (python 3.7)
$ pip install requests[security]@https://files.pythonhosted.org/packages/52/2c/514e4ac25da2b08ca5a464c50463682126385c4272c18193876e91f4bc38/requests-2.21.0.tar.gz
Requirement 'requests[security]@https://files.pythonhosted.org/packages/52/2c/514e4ac25da2b08ca5a464c50463682126385c4272c18193876e91f4bc38/requests-2.21.0.tar.gz' looks like a filename, but the file does not exist
Processing ./requests[security]@https:/files.pythonhosted.org/packages/52/2c/514e4ac25da2b08ca5a464c50463682126385c4272c18193876e91f4bc38/requests-2.21.0.tar.gz
Could not install packages due to an EnvironmentError: [Errno 2] No such file or directory: '/Users/andy/.virtualenvs/requests[security]@https:/files.pythonhosted.org/packages/52/2c/514e4ac25da2b08ca5a464c50463682126385c4272c18193876e91f4bc38/requests-2.21.0.tar.gz'
```

it works as expected in  `setup.py` though.

Reply:
It looks like there's lots of weird cases around PEP 508 URLs: https://github.com/pypa/pip/pull/5788#issuecomment-421849407 😕

Reply:
@pradyunsg updated with a test for the extras behavior 🤞